repository/src/lib.rs

```diff
@@ -188,46 +188,65 @@ impl<T: RawRepository> DistributedRepository<T> {
     /// It will leave only
     /// - the `finalized` branch
     /// - the `work` branch
-    /// - the `fp` branch.
+    /// - the `fp` branch
+    /// when `hard` is `true`,
     ///
-    /// and
+    /// and when `hard` is `false`,
     /// - the `p` branch
     /// - the `a-#` branches
     /// - the `b-#` branches
-    /// if only the branches are not outdated (branched from the last finalized commit).
-    pub async fn clean(&mut self) -> Result<(), Error> {
+    /// will be left as well
+    /// if only the branches have valid commit sequences
+    /// and are not outdated (branched from the last finalized commit).
+    pub async fn clean(&mut self, hard: bool) -> Result<(), Error> {
         let finalized_branch_commit_hash =
             self.raw.locate_branch(FINALIZED_BRANCH_NAME.into()).await?;
-
-        let branches = self.raw.list_branches().await?;
-
-        // delete outdated p branch, a-# branches, b-# branches
-        for branch in branches {
+        let branches = retrieve_local_branches(&self.raw).await?;
+        let last_header = self.get_last_finalized_block_header().await?;
```

Comment:
check integrity error

Reply:
`get_last_finalized_block_header` already handles integrity error itself. Should I fix this as well?